From 84ca7d811f9381ee744fee5d7e64fd8c1c1667c8 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Fri, 31 Aug 2018 14:02:17 +0200
Subject: [PATCH 39/44] ethtool: implement ACT_NWAY_RST message

Requests autonegotiation restart, equivalent of ETHTOOL_NWAY_RST.

If successful, send a notification in the same format.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 Documentation/networking/ethtool-netlink.txt | 15 +++-
 include/uapi/linux/ethtool_netlink.h         | 11 +++
 net/ethtool/Makefile                         |  3 +-
 net/ethtool/actions.c                        | 72 ++++++++++++++++++++
 net/ethtool/ioctl.c                          |  7 +-
 net/ethtool/netlink.c                        | 10 ++-
 net/ethtool/netlink.h                        |  9 +++
 7 files changed, 122 insertions(+), 5 deletions(-)
 create mode 100644 net/ethtool/actions.c

diff --git a/Documentation/networking/ethtool-netlink.txt b/Documentation/networking/ethtool-netlink.txt
index 19cd89ce0dfa..2cf912f3fb6d 100644
--- a/Documentation/networking/ethtool-netlink.txt
+++ b/Documentation/networking/ethtool-netlink.txt
@@ -131,6 +131,7 @@ List of message types
     ETHNL_CMD_SET_SETTINGS
     ETHNL_CMD_GET_PARAMS
     ETHNL_CMD_SET_PARAMS
+    ETHNL_CMD_ACT_NWAY_RST
 
 All constants use ETHNL_CMD_ prefix, usually followed by "GET", "SET" or "ACT"
 to indicate the type.
@@ -553,6 +554,18 @@ cannot be set by SET_PARAMS request.
 		- change configured FEC encodings
 
 
+ACT_NWAY_RST
+------------
+
+Request autonegotiation restart (equivalent of ETHTOOL_NWAY_RST).
+
+Request contents:
+
+    ETHTOOL_A_NWAYRST_DEV		(nested)	device identification
+
+If successful, a notification in the same format is sent.
+
+
 Request translation
 -------------------
 
@@ -570,7 +583,7 @@ ETHTOOL_GWOL			ETHNL_CMD_GET_SETTINGS
 ETHTOOL_SWOL			ETHNL_CMD_SET_SETTINGS
 ETHTOOL_GMSGLVL			ETHNL_CMD_GET_SETTINGS
 ETHTOOL_SMSGLVL			ETHNL_CMD_SET_SETTINGS
-ETHTOOL_NWAY_RST		n/a
+ETHTOOL_NWAY_RST		ETHNL_CMD_ACT_NWAY_RST
 ETHTOOL_GLINK			ETHNL_CMD_GET_SETTINGS
 ETHTOOL_GEEPROM			n/a
 ETHTOOL_SEEPROM			n/a
diff --git a/include/uapi/linux/ethtool_netlink.h b/include/uapi/linux/ethtool_netlink.h
index a0c7d90c971c..12415e73ac22 100644
--- a/include/uapi/linux/ethtool_netlink.h
+++ b/include/uapi/linux/ethtool_netlink.h
@@ -22,6 +22,7 @@ enum {
 	ETHNL_CMD_SET_SETTINGS,
 	ETHNL_CMD_GET_PARAMS,
 	ETHNL_CMD_SET_PARAMS,
+	ETHNL_CMD_ACT_NWAY_RST,
 
 	__ETHNL_CMD_CNT,
 	ETHNL_CMD_MAX = (__ETHNL_CMD_CNT - 1)
@@ -407,6 +408,16 @@ enum {
 	ETHTOOL_A_FEC_MAX = (__ETHTOOL_A_FEC_CNT - 1)
 };
 
+/* ACT_NWAY_RST */
+
+enum {
+	ETHTOOL_A_NWAYRST_UNSPEC,
+	ETHTOOL_A_NWAYRST_DEV,			/* nest - ETHTOOL_A_DEV_* */
+
+	__ETHTOOL_A_NWAYRST_CNT,
+	ETHTOOL_A_NWAYRST_MAX = (__ETHTOOL_A_NWAYRST_CNT - 1)
+};
+
 /* generic netlink info */
 #define ETHTOOL_GENL_NAME "ethtool"
 #define ETHTOOL_GENL_VERSION 1
diff --git a/net/ethtool/Makefile b/net/ethtool/Makefile
index 95e93c153cf6..b3e88cf0f4c0 100644
--- a/net/ethtool/Makefile
+++ b/net/ethtool/Makefile
@@ -4,4 +4,5 @@ obj-y				+= ioctl.o common.o
 
 obj-$(CONFIG_ETHTOOL_NETLINK)	+= ethtool_nl.o
 
-ethtool_nl-y	:= netlink.o bitset.o strset.o info.o settings.o params.o
+ethtool_nl-y	:= netlink.o bitset.o strset.o info.o settings.o params.o \
+		   actions.o
diff --git a/net/ethtool/actions.c b/net/ethtool/actions.c
new file mode 100644
index 000000000000..2f3f9e85016d
--- /dev/null
+++ b/net/ethtool/actions.c
@@ -0,0 +1,72 @@
+/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
+
+#include "netlink.h"
+#include "common.h"
+
+/* ACT_NWAY_RST */
+
+static const struct nla_policy nwayrst_policy[ETHTOOL_A_NWAYRST_MAX + 1] = {
+	[ETHTOOL_A_NWAYRST_UNSPEC]	= { .type = NLA_REJECT },
+	[ETHTOOL_A_NWAYRST_DEV]		= { .type = NLA_NESTED },
+};
+
+void ethnl_nwayrst_notify(struct net_device *dev,
+                          struct netlink_ext_ack *extack, unsigned int cmd,
+                          u32 req_mask, const void *data)
+{
+	struct sk_buff *skb;
+	void *msg_payload;
+	int msg_len;
+	int ret;
+
+	msg_len = dev_ident_size();
+	skb = genlmsg_new(msg_len, GFP_KERNEL);
+	if (!skb)
+		return;
+	msg_payload = ethnl_bcastmsg_put(skb, ETHNL_CMD_ACT_NWAY_RST);
+	if (!msg_payload)
+		goto err_skb;
+
+	ret = ethnl_fill_dev(skb, dev, ETHTOOL_A_NWAYRST_DEV);
+	if (ret < 0)
+		goto err_skb;
+	genlmsg_end(skb, msg_payload);
+	ethnl_multicast(skb, dev);
+	return;
+
+err_skb:
+	nlmsg_free(skb);
+}
+
+int ethnl_act_nway_rst(struct sk_buff *skb, struct genl_info *info)
+{
+	struct nlattr *tb[ETHTOOL_A_NWAYRST_MAX + 1];
+	struct net_device *dev;
+	int ret;
+
+	ret = nlmsg_parse(info->nlhdr, GENL_HDRLEN, tb, ETHTOOL_A_NWAYRST_MAX,
+			  nwayrst_policy, info->extack);
+	if (ret < 0)
+		return ret;
+	dev = ethnl_dev_get(info, tb[ETHTOOL_A_NWAYRST_DEV]);
+	if (IS_ERR(dev))
+		return PTR_ERR(dev);
+	ret = -EOPNOTSUPP;
+	if (!dev->ethtool_ops->nway_reset)
+		goto out_dev;
+
+	rtnl_lock();
+	ret = ethnl_before_ops(dev);
+	if (ret < 0)
+		goto out_rtnl;
+	ret = dev->ethtool_ops->nway_reset(dev);
+	ethnl_after_ops(dev);
+	if (ret == 0)
+		ethtool_notify(dev, NULL, ETHNL_CMD_ACT_NWAY_RST, 0, NULL);
+
+out_rtnl:
+	rtnl_unlock();
+out_dev:
+	dev_put(dev);
+	return ret;
+}
diff --git a/net/ethtool/ioctl.c b/net/ethtool/ioctl.c
index d14d8e7edf55..d34af84c1adf 100644
--- a/net/ethtool/ioctl.c
+++ b/net/ethtool/ioctl.c
@@ -1282,10 +1282,15 @@ static int ethtool_set_eee(struct net_device *dev, char __user *useraddr)
 
 static int ethtool_nway_reset(struct net_device *dev)
 {
+	int ret;
+
 	if (!dev->ethtool_ops->nway_reset)
 		return -EOPNOTSUPP;
+	ret = dev->ethtool_ops->nway_reset(dev);
+	if (ret == 0)
+		ethtool_notify(dev, NULL, ETHNL_CMD_ACT_NWAY_RST, 0, NULL);
 
-	return dev->ethtool_ops->nway_reset(dev);
+	return ret;
 }
 
 static int ethtool_get_link(struct net_device *dev, char __user *useraddr)
diff --git a/net/ethtool/netlink.c b/net/ethtool/netlink.c
index 0f6def083d73..dcf4f16210fa 100644
--- a/net/ethtool/netlink.c
+++ b/net/ethtool/netlink.c
@@ -240,13 +240,13 @@ struct sk_buff *ethnl_reply_init(size_t payload, struct net_device *dev, u8 cmd,
 	return NULL;
 }
 
-static void *ethnl_bcastmsg_put(struct sk_buff *skb, u8 cmd)
+void *ethnl_bcastmsg_put(struct sk_buff *skb, u8 cmd)
 {
 	return genlmsg_put(skb, 0, ++ethnl_bcast_seq, &ethtool_genl_family, 0,
 			   cmd);
 }
 
-static int ethnl_multicast(struct sk_buff *skb, struct net_device *dev)
+int ethnl_multicast(struct sk_buff *skb, struct net_device *dev)
 {
 	return genlmsg_multicast_netns(&ethtool_genl_family, dev_net(dev), skb,
 				       0, ETHNL_MCGRP_MONITOR, GFP_KERNEL);
@@ -583,6 +583,7 @@ typedef void (*ethnl_notify_handler_t)(struct net_device *dev,
 static const ethnl_notify_handler_t ethnl_notify_handlers[] = {
 	[ETHNL_CMD_SET_SETTINGS]	= ethnl_std_notify,
 	[ETHNL_CMD_SET_PARAMS]		= ethnl_std_notify,
+	[ETHNL_CMD_ACT_NWAY_RST]	= ethnl_nwayrst_notify,
 };
 
 void ethtool_notify(struct net_device *dev, struct netlink_ext_ack *extack,
@@ -714,6 +715,11 @@ static const struct genl_ops ethtool_genl_ops[] = {
 		.flags	= GENL_UNS_ADMIN_PERM,
 		.doit	= ethnl_set_params,
 	},
+	{
+		.cmd	= ETHNL_CMD_ACT_NWAY_RST,
+		.flags	= GENL_UNS_ADMIN_PERM,
+		.doit	= ethnl_act_nway_rst,
+	},
 };
 
 static const struct genl_multicast_group ethtool_nl_mcgrps[] = {
diff --git a/net/ethtool/netlink.h b/net/ethtool/netlink.h
index ba71dc367869..52b4d692fca2 100644
--- a/net/ethtool/netlink.h
+++ b/net/ethtool/netlink.h
@@ -24,6 +24,8 @@ int ethnl_fill_dev(struct sk_buff *msg, struct net_device *dev, u16 attrtype);
 struct sk_buff *ethnl_reply_init(size_t payload, struct net_device *dev, u8 cmd,
 				 u16 dev_attrtype, struct genl_info *info,
 				 void **ehdrp);
+void *ethnl_bcastmsg_put(struct sk_buff *skb, u8 cmd);
+int ethnl_multicast(struct sk_buff *skb, struct net_device *dev);
 
 #if BITS_PER_LONG == 64 && defined(__BIG_ENDIAN)
 void ethnl_bitmap_to_u32(unsigned long *bitmap, unsigned int nwords);
@@ -261,5 +263,12 @@ extern const struct get_request_ops params_request_ops;
 
 int ethnl_set_settings(struct sk_buff *skb, struct genl_info *info);
 int ethnl_set_params(struct sk_buff *skb, struct genl_info *info);
+int ethnl_act_nway_rst(struct sk_buff *skb, struct genl_info *info);
+
+/* notify handlers */
+
+void ethnl_nwayrst_notify(struct net_device *dev,
+			  struct netlink_ext_ack *extack, unsigned int cmd,
+			  u32 req_mask, const void *data);
 
 #endif /* _NET_ETHTOOL_NETLINK_H */
-- 
2.21.0

